### PR TITLE
feat: Implement Quota Limiting Suspension Strategy

### PR DESCRIPTION
**Pull request by xrdt**
## Problem

RFC: https://github.com/PostHog/product-internal/pull/564/files

Billing changes: https://github.com/PostHog/billing/pull/468

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/migrations/0391_organization_is_hipaa_and_more.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("posthog", "0390_personalapikey_scopes"),
+    ]
+
+    operations = [
```
Comment (xrdt) [inline posthog/migrations/0391_organization_is_hipaa_and_more.py:11]: Am going to leave this hipaa change here and move the "trusted_customer_scores" migration into a separate PR

File: posthog/migrations/0392_organization_is_hipaa.py
Comment (raquelmsmith) [inline posthog/migrations/0392_organization_is_hipaa.py:1]: I think we generally try to only have one migration per PR, I'd delete these files and re-run migrations.

File: posthog/models/organization.py [hunk 2]
```diff
@@ -125,6 +121,10 @@ class PluginsAccessLevel(models.IntegerChoices):
     for_internal_metrics: models.BooleanField = models.BooleanField(default=False)
     is_member_join_email_enabled: models.BooleanField = models.BooleanField(default=True)
     enforce_2fa: models.BooleanField = models.BooleanField(null=True, blank=True)
+    # Scoring mechanism defined here: https://github.com/PostHog/product-internal/pull/564
+    trusted_customer_scores: models.JSONField = models.JSONField(default=dict, null=True, blank=True)
+    trusted_customer_score_overrides: models.JSONField = models.JSONField(default=dict, null=True, blank=True)
```
Comment (raquelmsmith) [inline posthog/models/organization.py:126]: What is the purpose of this field?

Comment (xrdt) [reply posthog/models/organization.py:126]: This seemed like the easiest way to support manual score overrides (e.g. through django admin) without interfering with the automatic score calculation mechanism. We store two dictionaries and merge them at the point of use in quota_limiting.

File: ee/billing/quota_limiting.py [hunk 3]
```diff
@@ -87,42 +101,174 @@ def org_quota_limited_until(organization: Organization, resource: QuotaResource)
...
-        if quota_limited_until:
-            add_limited_team_tokens(resource, {x: quota_limited_until for x in team_attributes})
+        result = org_quota_limited_until(organization, resource, previously_quota_limited_team_tokens)
+
+        if result:
+            quota_limited_until = result.get("quota_limited_until")
+            limiting_suspended_until = result.get("quota_limiting_suspended_until")
+            needs_save = result.get("needs_save")
+            if needs_save:
+                organization.save(update_fields=["usage"])
```
[the remark this answers is not in the document]
Comment (xrdt) [reply ee/billing/quota_limiting.py:252]: I was trying to minimize the number of functions in which we call `.save()` ... Maybe it will come in handy in the future, maybe it won't, I'm happy to move `.save()` into `org_quota_limited_until`.

File: posthog/models/organization.py (hunk 2 above)
Comment (raquelmsmith) [reply posthog/models/organization.py:126]: I think a single map is probably better and achieves the same result - it's only incremented if they meet the criteria for a higher score, and is never programmatically decremented. Someone (eg CS) can go in and set it manually if they want. If they manually downgrade someone, then it won't automatically upgrade again if we don't have the `highest_paid_bill` in here and just go based on the current bill.

Comment (raquelmsmith) [reply posthog/models/organization.py:126]: So the situation is - a customer is paying a bunch of money so would have a higher score, but we don't like them or they commonly make implementation errors so we manually downgrade them to drop when they hit their quota. 

IDK how often that will happen. This is only relevant for people who are on the free plan or have billing limits set, and are prone to hitting those limits.

Comment (xrdt) [reply posthog/models/organization.py:126]: Yes, that's true, but if we maintain an automatic score recalculation system (and I think we should do this in billing/customer any time a subscription changes or a new invoice is completd), their score would simply be re-upped to what it was. 

For example, if they're on level 7 but get manually overwritten to 0, then upgrade to the teams plan, the automatic scoring mechanism would mark them as 10.

**Comment by xrdt** [comment]
Remember that customer_trust_scores can be blank / null

File: posthog/management/commands/update_billing_quotas.py [hunk 4]
```diff
@@ -15,16 +15,23 @@ def add_arguments(self, parser):
...
         else:
-            print(f"{len(results['events'])} orgs rate limited for events")  # noqa T201
-            print(f"{len(results['recordings'])} orgs rate limited for recordings")  # noqa T201
+            print(f"{len(quota_limited_orgs['events'])} orgs rate limited for events")  # noqa T201
+            print(  # noqa T201
+                f"{len(quota_limiting_suspended_orgs['events'])} orgs rate rate limiting suspended for events"  # noqa T201
+            )  # noqa T201
+            print(f"{len(quota_limited_orgs['recordings'])} orgs rate limited for recordings")  # noqa T201
+            print(  # noqa T201
+                f"{len(quota_limiting_suspended_orgs['recordings'])} orgs rate rate limiting suspended for recordings"  # noqa T201
```
Comment (xrdt) [inline posthog/management/commands/update_billing_quotas.py:35]:
```suggestion
                f"{len(quota_limiting_suspended_orgs['recordings'])} orgs rate limiting suspended for recordings"  # noqa T201
```

**Comment by xrdt** [comment]
@raquelmsmith fields updated to match billing.

File: ee/billing/quota_limiting.py [hunk 5]
```diff
@@ -86,43 +101,167 @@ def org_quota_limited_until(organization: Organization, resource: QuotaResource)
...
+    if posthoganalytics.feature_enabled(
         QUOTA_LIMIT_DATA_RETENTION_FLAG,
         organization.id,
         groups={"organization": str(organization.id)},
         group_properties={"organization": {"id": str(organization.id)}},
     ):
-        # Don't drop data for this org but record that they __would have__ been
-        # limited.
+        # Only suspend quota limiting for teams that are not already being limited.
+        # Don't drop data for this org but record that they would have been limited.
```
Comment (raquelmsmith) [inline ee/billing/quota_limiting.py:141]:
```suggestion
        # Don't drop data for this org but record that they would have been limited.
```

File: ee/billing/test/test_quota_limiting.py [hunk 6]
```diff
@@ -54,54 +62,93 @@ def test_dont_quota_limit_feature_flag_enabled(self, patch_feature_enabled, patc
...
+            groups={"organization": org_id},
+            group_properties={"organization": {"id": org_id}},
+        )
+        patch_capture.assert_not_called()
 
-        assert self.redis_client.zrange(f"{QUOTA_LIMITER_CACHE_KEY}events", 0, -1) == []
-        assert self.redis_client.zrange(f"{QUOTA_LIMITER_CACHE_KEY}recordings", 0, -1) == []
-        assert self.redis_client.zrange(f"{QUOTA_LIMITER_CACHE_KEY}rows_synced", 0, -1) == []
+        # Feature flag is on but we only suspend limiting for orgs that were not previously limited. This org should still be in the set.
+        assert quota_limited_orgs["events"] == {org_id: 1612137599}
```
Comment (raquelmsmith) [inline ee/billing/test/test_quota_limiting.py:111]: is this org id going to change ever?

File: posthog/migrations/0394_organization_customer_trust_scores_and_more.py
Comment (raquelmsmith) [inline posthog/migrations/0394_organization_customer_trust_scores_and_more.py:1]: For ease of revert-ability, I wonder if we should put this migration in as a separate PR. The field is not going to cause issues, but the logic might, and if we have an issue like last time (hopefully we won't!) then this migration would make a revert untenable, and therefore slower to recover from.

**Review by raquelmsmith — APPROVED** [review]
🙌

**Comment by xrdt** [comment]
Ok @raquelmsmith! Migrations are pulled out into a separate pr and the other comments are addressed so will go fix up the backend.
